apps/daimo-mobile/src/logic/systemContacts.ts

```diff
@@ -0,0 +1,149 @@
+    assert(!!contactsPermission, "Contacts permission not loaded");
+
+    if (!contactsPermission.granted) {
+      await requestContactsAccess(contactsPermission.canAskAgain);
+    }
+    Contacts.getPermissionsAsync().then(setContactsPermission);
+  };
+
+  return { permission: contactsPermission, ask };
+}
+
+// Take user through a two-step flow to ask for contacts access.
+// Promise resolves when user has finished the flow (either granting
+// or denying).
+function requestContactsAccess(canAskAgain: boolean) {
+  // Step 1: Ask with our own prompt, re-assuring user.
+  return new Promise<void>((resolve) => {
+    Alert.alert(
+      "Daimo Would Like to Access Your Contacts",
+      "Your contacts are accessed locally to find and pay friends with Daimo.\n\nThey are never uploaded to our servers.",
+      [
+        {
+          text: `Continue`,
+          onPress: () => {
+            if (canAskAgain) askSystem(resolve);
+            else askOpenSettings(resolve);
+          },
+        },
+        {
+          text: "Don't Allow",
+          style: "cancel",
+          onPress: () => {
+            resolve();
+          },
+        },
+      ]
+    );
+  });
+}
+
+function askSystem(resolve: (value: void) => void) {
+  // Step 2: If user says yes to our own prompt, ask with the system prompt
+  Contacts.requestPermissionsAsync().then(() => {
+    resolve();
+  });
+}
+
+async function askOpenSettings(resolve: (value: void) => void) {
+  // Step 2: If user says yes to our own prompt but previously denied system
+  // prompt, ask them to open settings.
+  Alert.alert(
+    "Enable access in Settings",
+    "Visit Settings > Daimo and enable contacts.",
+    [
+      {
+        text: "Continue",
+        onPress: () => {
+          Linking.openSettings().then(async () => {
+            // On iOS, the app is reset by the OS when access is changed from
+            // settings, so this isn't triggered. On Android, we wait a
+            // few seconds to let user open settings, then resolve.
+            await new Promise((f) => setTimeout(f, 5000));
+            resolve();
+          });
+        },
+      },
+      {
+        text: "Cancel",
+        style: "cancel",
+        onPress: () => {
+          resolve();
+        },
+      },
+    ]
+  );
+}
+
+export function useSystemContactsSearch(
+  prefix: string,
+  enabled: boolean
+): SystemContact[] {
+  const [recipients, setRecipients] = useState<SystemContact[]>([]);
+
+  useEffect(() => {
+    const search = async (nameSearchTerm: string) => {
+      const { data } = await Contacts.getContactsAsync({
+        // TODO: add images?
+        fields: [Contacts.Fields.PhoneNumbers, Contacts.Fields.Emails],
+        name: nameSearchTerm,
+        pageSize: 5,
+      });
+
+      const matches = [] as SystemContact[];
+      for (const contact of data) {
+        const name = contact.name;
+        if (name == null || !name.toLowerCase().includes(nameSearchTerm))
```

Comment:
what if i type in `jmarc` and I have a contact Jon Marcus / jmarc3@gmail.com

do we get extraneous results from system search that we have to filter out? (if so, recommend matching either name or email)

Reply:
> do we get extraneous results from system search that we have to filter out?

no, only name matches are returned, from doc: `Get all contacts whose name contains the provided string (not case-sensitive).` This line was leftover from a previous experiment, removing 👍

Sadly, the `expo-contacts` search API only exposes this easy handler and doing full load and search is too slow (took ~2 seconds on my iPhone with 490 contacts).

In future if we want to add search by email / phone number we'll need to make our own local sqlite DB and index the contacts.